### ISSUE TITLE
Add parser for mbox

### ISSUE THREAD

**Issue by joachimmetz** [log2timeline/plaso#2151]
Taken from the Plaso - Roadmap and Assignment

**Comment by stefanomart** [comment]
Hi, I would like to develop this parser. Do you already have any idea about how it should be?
I have some doubts about how to validate a MBox file: I currently check on the file extension and how it starts ("From a ...").

**Comment by Onager** [comment]
Hey @stefanomart - this approach to validating an mbox file sounds OK, but I'd recommend dropping the file extension check.

Given that mbox is a text file format, I recommend create a subclass of the text parser - probably PyParsingMultilineTextParser

**Comment by Onager** [comment]
I also note there are several mbox-relateds formats: https://en.wikipedia.org/wiki/Mbox

**Comment by Onager** [comment]
https://docs.python.org/2/library/mailbox.html#mbox looks helpful for handling mboxo files

**Comment by stefanomart** [comment]
Thanks for the reply. I'm using the mailbox module to get any information. I noticed that an email can have more than one field with the same name (like the field _received_). Using the subclass **mbox** I can get only the first occurrence of a field that is more than once in a same mail. Do you have some advice?

**Comment by stefanomart** [comment]
@Onager 
Hi, I'm using the mailbox library to parser some mbox file. The class mailbox.mbox wants as parameter the path of the file, not the file like (file_object in plaso). Would this be a problem?
Thanks in advance.

**Comment by joachimmetz** [comment]
> Would this be a problem?

Yes, plaso uses dfVFS to read files directly from storage media images and devices. Your parser will need to be able to work with a file-like object.

There are workarounds like copying the file to a temporary location, however this is no recommended.

**Comment by joachimmetz** [comment]
To be verbose, please also read https://github.com/log2timeline/plaso/wiki/Adding-a-new-dependency.

**Comment by stefanomart** [comment]
> > Would this be a problem?
> 
> Yes, plaso uses dfVFS to read files directly from storage media images and devices. Your parser will need to be able to work with a file-like object.
> 
> There are workarounds like copying the file to a temporary location, however this is no recommended.

Would it be possible to use the mailbox library partially modified by me? In order to have as a parameter a file-like object and not a path.

**Comment by Onager** [comment]
mbox is in the standard library, so getting an upstream change in might be challenging. It might be possible to just add a subclass of the standard library mbox that takes a file-like object, and use that in your parser: https://github.com/python/cpython/blob/3.6/Lib/mailbox.py

**Comment by joachimmetz** [comment]
Various standard library modules of Python are maintained as stand-alone projects. Not sure but the API of this project looks very similar https://pypi.org/project/mailbox/

**Comment by stefanomart** [comment]
Thank you for the replies.
I have already made the parser for mbox using the mailbox library. I could add the logic of this library into plaso, eliminating all the irrelevant and potentially dangerous parts (like the writing to file). In this way you don't risk adding code with potential errors, and I could continue the work I started.
Moreover, the mailbox library is in use since 1994, so we can be sure it's battle-tested and handles all the corner cases correctly.

**Comment by joachimmetz** [comment]
>  could add the logic of this library into plaso, eliminating all the irrelevant and potentially dangerous parts (like the writing to file).

Sounds like the more feasible option then ;)

> Moreover, the mailbox library is in use since 1994, so we can be sure it's battle-tested and handles all the corner cases correctly.

I'm sceptical here, about handling "all" edge (corner) cases, but likely a fair bit. It is nice that it's part of the core, which means no additional changes for deployment.

**Comment by stefanomart** [comment]
I'm using part of mailbox's logic and I've a problem with the file object (dfvfs.FileIO): because this class doesn't implement the method _readline()_ , I created it but it's way slower than File's _readline()_. I also tried to implement the method based on the logic of text_file's _readline()_, but the results are not good anyway.
Why dfvfs.FileIO doesn't have _readline()_? By the way, do you have any suggestions?

**Comment by joachimmetz** [comment]
> Why dfvfs.FileIO doesn't have readline()?

Because it is not needed most of the time and readline is not a [file IO primitive](https://www.gnu.org/software/libc/manual/html_node/I_002fO-Primitives.html). 

> I also tried to implement the method based on the logic of text_file's readline(), but the results are not good anyway.

based on this comment I assume you tried https://github.com/log2timeline/dfvfs/blob/master/dfvfs/helpers/text_file.py? could you be more specific in terms what way the results are not good?

**Comment by Onager** [comment]
I wouldn't worry too much about that stefanomart@, Plaso's run time is pretty long, and the extra time for I/O is the price we pay for having a dfvfs to handle all the different file systems/file storage formats.
